numpy/linalg/linalg.py

```diff
@@ -2317,8 +2317,10 @@ def lstsq(a, b, rcond="warn"):
         # we probably should squeeze resids too, but we can't
         # without breaking compatibility.
 
+    # Since matrices are stacked all the ranks will be same.
+    rank_val = rank[0] if a.ndim > 2 else rank
```

Comment:
I don't think this claim is true

Reply:
I was having a similar feeling when I made this change. Should we take the maximum rank?

Reply:
We should return the array of ranks.

Reply:
For matrices which have `m <= n`, the rank value doesn't matter, since the residue array will be returned empty (no matter if the matrices are stacked or not, the results are same for both the cases). For matrices with `m > n`, the value of rank will matter in which case `rank <= n` and if it is strictly less than `n`, then again the residues will be empty. I tested for some cases and found out that either they are `0.` (for m <= n) or `nan` (for rank < n). I am adding some examples below for more clarity. 

P.S. Anyways what about always returning the residues as it is for stacked matrices(for simple matrix inputs we can keep the behavior same). The reason why I am saying so is that, stacked matrices are not supported yet so whatever we do to support them will not be affecting the current APIs.

cc: @asmeurer @rgommers @leofang @charris What do you say on the above thoughts?

Reply:
**Code**

```python
np.random.seed(0)

a1, b1 = np.random.normal(size=(3, 2)), np.random.normal(size=(3, 2))
a2, b2 = np.random.normal(size=(3, 2)), np.random.normal(size=(3, 2))
a3, b3 = np.random.normal(size=(3, 2)), np.random.normal(size=(3, 2))
a4, b4 = np.asarray([[1, -1], [0, 0], [0, 0]]), np.asarray([[3, 2], [2, -3], [-1, -4]]) # a4 is rank 1

a = np.asarray([[a1, a2], [a3, a4]])
b = np.asarray([[b1, b2], [b3, b4]])
```

**Output**
```bash
Stacked Outputs
===============

(array([[[[ 2.50022830e-01,  4.00171540e-01],
         [-4.24355386e-02, -1.61563323e-01]],

        [[ 2.97027623e-03, -4.57468415e-01],
         [-5.89018640e+00,  1.40474842e+00]]],


       [[[-4.99964661e-02,  1.40774201e-01],
         [-1.44630369e-01,  4.62672012e-02]],

        [[ 1.50000000e+00,  1.00000000e+00],
         [-1.50000000e+00, -1.00000000e+00]]]]), array([[[0.47337501, 1.07485169],
        [1.5239536 , 0.66133179]],

       [[0.8396116 , 3.94686079],
        [       nan,        nan]]]), array([[2, 2],
       [2, 1]], dtype=int32), array([[[2.85217354, 2.35785662],
        [1.73493218, 0.40829999]],

       [[2.79671766, 1.99772241],
        [1.41421356, 0.        ]]]))
===============

Outputs for a1
==============

(array([[ 0.25002283,  0.40017154],
       [-0.04243554, -0.16156332]]), array([0.47337501, 1.07485169]), 2, array([2.85217354, 2.35785662]))
==============

Outputs for a2
==============

(array([[ 2.97027623e-03, -4.57468415e-01],
       [-5.89018640e+00,  1.40474842e+00]]), array([1.5239536 , 0.66133179]), 2, array([1.73493218, 0.40829999]))
==============

Outputs for a3
==============

(array([[-0.04999647,  0.1407742 ],
       [-0.14463037,  0.0462672 ]]), array([0.8396116 , 3.94686079]), 2, array([2.79671766, 1.99772241]))
==============

Outputs for a4
==============

(array([[ 1.5,  1. ],
       [-1.5, -1. ]]), array([], dtype=float64), 1, array([1.41421356, 0.        ]))
==============
```